internal/kibana/role.go

```diff
@@ -0,0 +1,506 @@
+		},
+	}
+
+	return &schema.Resource{
+		Description: "Creates a Kibana role. See, https://www.elastic.co/guide/en/kibana/master/role-management-api-put.html",
+
+		CreateContext: resourceRoleUpsert,
+		UpdateContext: resourceRoleUpsert,
+		ReadContext:   resourceRoleRead,
+		DeleteContext: resourceRoleDelete,
+
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: apikeySchema,
+	}
+}
+
+func resourceRoleUpsert(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, diags := clients.NewApiClient(d, meta)
+	if diags.HasError() {
+		return diags
+	}
+
+	kibana, err := client.GetKibanaClient()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	queryParams := ""
+	if d.IsNewResource() {
+		queryParams = "?createOnly=true"
+	}
+	kibanaRole := kbapi.KibanaRole{
+		Name:          fmt.Sprintf("%s%s", d.Get("name").(string), queryParams),
+		Kibana:        []kbapi.KibanaRoleKibana{},
+		Elasticsearch: &kbapi.KibanaRoleElasticsearch{},
+	}
+
+	if v, ok := d.GetOk("kibana"); ok {
+		definedKibanaConfigs := v.(*schema.Set)
+		kibanaRole.Kibana = make([]kbapi.KibanaRoleKibana, definedKibanaConfigs.Len())
+		for i, item := range definedKibanaConfigs.List() {
+			each := item.(map[string]interface{})
+			_config := kbapi.KibanaRoleKibana{
+				Base:    []string{},
+				Feature: map[string][]string{},
+			}
+
+			if basePrivileges, ok := each["base"].(*schema.Set); ok && basePrivileges.Len() > 0 {
+				if _features, ok := each["feature"].(*schema.Set); ok && _features.Len() > 0 {
+					return diag.Errorf("Only one of the `feature` or `base` privileges allowed!")
+				}
+				_config.Base = make([]string, basePrivileges.Len())
+				for i, name := range basePrivileges.List() {
+					_config.Base[i] = name.(string)
+				}
+			} else if kibanaFeatures, ok := each["feature"].(*schema.Set); ok && kibanaFeatures.Len() > 0 {
+				for _, item := range kibanaFeatures.List() {
+					featureData := item.(map[string]interface{})
+					featurePrivileges := featureData["privileges"].(*schema.Set)
+					_features := make([]string, featurePrivileges.Len())
+					for i, f := range featurePrivileges.List() {
+						_features[i] = f.(string)
+					}
+					_config.Feature[featureData["name"].(string)] = _features
+				}
+			} else {
+				return diag.Errorf("Either on of the `feature` or `base` privileges must be set for kibana role!")
+			}
+
+			if roleSpaces, ok := each["spaces"].(*schema.Set); ok && roleSpaces.Len() > 0 {
+				_config.Spaces = make([]string, roleSpaces.Len())
+				for i, name := range roleSpaces.List() {
+					_config.Spaces[i] = name.(string)
+				}
+			}
+			kibanaRole.Kibana[i] = _config
+		}
+	}
+
+	if v, ok := d.GetOk("elasticsearch"); ok {
+		if definedElasicConfigs := v.(*schema.Set); definedElasicConfigs.Len() > 0 {
+			userElasitcConfig := definedElasicConfigs.List()[0].(map[string]interface{})
+			if v, ok := userElasitcConfig["cluster"]; ok {
```

Comment:
There's a lot going on in here. It would be nicer to break some of these out to `expand<Thing>` functions.